tools/clang/lib/SPIRV/SPIRVEmitter.cpp

```diff
@@ -2212,29 +2220,82 @@ uint32_t SPIRVEmitter::processIntrinsicDot(const CallExpr *callExpr) {
 
 uint32_t SPIRVEmitter::processIntrinsicAllOrAny(const CallExpr *callExpr,
                                                 spv::Op spvOp) {
-  const uint32_t returnType = typeTranslator.translateType(callExpr->getType());
-
   // 'all' and 'any' take only 1 parameter.
   assert(callExpr->getNumArgs() == 1u);
+  const QualType returnType = callExpr->getType();
+  const uint32_t returnTypeId = typeTranslator.translateType(returnType);
   const Expr *arg = callExpr->getArg(0);
   const QualType argType = arg->getType();
 
-  if (hlsl::IsHLSLMatType(argType)) {
-    emitError("'all' and 'any' do not support matrix arguments yet.");
-    return 0;
+  // Handle scalars, vectors of size 1, and 1x1 matrices as arguments.
+  // Optimization: can directly cast them to boolean. No need for OpAny/OpAll.
+  {
+    if (argType->isBooleanType() || argType->isFloatingType() ||
+        argType->isIntegerType() || TypeTranslator::isVec1Type(argType) ||
+        TypeTranslator::is1x1Matrix(argType))
+      return castToBool(doExpr(arg), argType, returnType);
   }
 
-  bool isSpirvAcceptableVecType =
-      hlsl::IsHLSLVecType(argType) && hlsl::GetHLSLVecSize(argType) > 1;
-  if (!isSpirvAcceptableVecType) {
-    // For a scalar or vector of 1 scalar, we can simply cast to boolean.
-    return castToBool(doExpr(arg), arg->getType(), callExpr->getType());
-  } else {
-    // First cast the vector to a vector of booleans, then use OpAll
-    uint32_t boolVecId =
-        castToBool(doExpr(arg), arg->getType(), callExpr->getType());
-    return theBuilder.createUnaryOp(spvOp, returnType, boolVecId);
+  // Handle vectors larger than 1, Mx1 matrices, and 1xN matrices as arguments.
+  // Cast the vector to a boolean vector, then run OpAny/OpAll on it.
+  {
+    QualType elemType = {};
+    uint32_t size = 0;
+    if (TypeTranslator::isVectorType(argType, &elemType, &size) ||
+        TypeTranslator::isMx1Or1xNMatrix(argType, &elemType, &size)) {
+      const QualType castToBoolType =
+          astContext.getExtVectorType(returnType, size);
+      uint32_t castedToBoolId =
+          castToBool(doExpr(arg), argType, castToBoolType);
+      return theBuilder.createUnaryOp(spvOp, returnTypeId, castedToBoolId);
+    }
+  }
+
+  // Handle MxN matrices as arguments.
+  {
+    if (!TypeTranslator::isSpirvAcceptableMatrixType(argType)) {
+      emitError("'all' and 'any' currently do not take non-floating point "
+                "matrices as argument.");
+      return 0;
+    }
+
+    QualType elemType = {};
+    uint32_t matRowCount = 0, matColCount = 0;
+    if (TypeTranslator::isMxNMatrix(argType, &elemType, &matRowCount,
```

Comment:
Can move this `isMxNMatrix` call to the beginning of this compound statement so that we don't need to call `isSpirvAcceptableMatrixType`.